Immense.RemoteControl.Server/Services/DesktopHubSessionCache.cs

```diff
@@ -26,61 +27,86 @@ public interface IDesktopHubSessionCache
+        lock (_sessionsLock)
         {
-            v.Dispose();
-            return session;
-        });
+            return AddOrUpdate(sessionId, session, (k, v) =>
+            {
+                v.Dispose();
+                return session;
+            });
+        }
     }
 
     public RemoteControlSession AddOrUpdate(
-        string sessionId, 
-        RemoteControlSession session, 
+        string sessionId,
+        RemoteControlSession session,
         Func<string, RemoteControlSession, RemoteControlSession> updateFactory)
     {
-        return _sessions.AddOrUpdate(sessionId, session, updateFactory);
-    }
+        lock (_sessionsLock)
+        {
+            if (_sessions.ContainsKey(sessionId))
+            {
+                var newValue = updateFactory(sessionId, _sessions[sessionId]);
+                _sessions[sessionId] = newValue;
+                return newValue;
+            }
 
-    public RemoteControlSession GetOrAdd(string sessionId, Func<string, RemoteControlSession> valueFactory)
-    {
-        return _sessions.GetOrAdd(sessionId, valueFactory);
+            _sessions[sessionId] = session;
+            NotifySessionAdded(session);
+            return session;
+        }
     }
 
-    public void Remove(string sessionId)
+    public RemoteControlSession GetOrAdd(string sessionId, Func<string, RemoteControlSession> valueFactory)
     {
-        if (_sessions.TryRemove(sessionId, out var session))
+        lock (_sessionsLock)
         {
-            session.Dispose();
+            return _sessions.GetOrAdd(sessionId, (key) =>
+            {
+                var session = valueFactory(key);
+                NotifySessionAdded(session);
+                return session;
+            });
         }
     }
 
     public Task RemoveExpiredSessions()
     {
-        foreach (var session in _sessions)
+        lock (_sessionsLock)
         {
-            if (session.Value.Mode is RemoteControlMode.Unattended or RemoteControlMode.Unknown &&
-                !session.Value.ViewerList.Any() &&
-                session.Value.Created < _systemTime.Now.AddMinutes(-1))
+            foreach (var session in _sessions)
             {
-                _logger.LogWarning("Removing expired session: {session}", JsonSerializer.Serialize(session.Value));
-                if (_sessions.TryRemove(session.Key, out var expiredSession))
+                if (session.Value.Mode is RemoteControlMode.Unattended or RemoteControlMode.Unknown &&
+                    !session.Value.ViewerList.Any() &&
+                    session.Value.Created < _systemTime.Now.AddMinutes(-1))
                 {
-                    expiredSession.Dispose();
+                    _logger.LogWarning("Removing expired session: {session}", JsonSerializer.Serialize(session.Value));
+                    if (_sessions.TryRemove(session.Key, out var expiredSession))
+                    {
+                        NotifySessionRemoved(expiredSession);
```

Comment:
It might be a good idea to implement an observable concurrent dictionary so that we don't have to remember to fire these events manually in the right places

Reply:
But then I'd have to remember to fire events in the right places within the `ObservableConcurrentDictionary`.  😆 

But seriously, if I have to do this more than once, then yeah, I'll make a re-usable class.